larpandoracontent/LArThreeDReco/LArCosmicRay/NViewDeltaRayMatchingAlgorithm.cc

```diff
@@ -0,0 +1,1105 @@
+
+    // Catch if delta ray has travelled along muon
+    if ((static_cast<float>(collectedHits.size()) / pMuonCluster->GetNCaloHits()) > m_maxCosmicRayHitFraction)
+        return STATUS_CODE_NOT_FOUND;
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template<typename T>
+StatusCode NViewDeltaRayMatchingAlgorithm<T>::ParameteriseMuon(const ParticleFlowObject *const pParentMuon, const Cluster *const pDeltaRayCluster,
+    CartesianVector &positionOnMuon, CartesianVector &muonDirection) const
+{
+    CaloHitList deltaRayHitList;
+    pDeltaRayCluster->GetOrderedCaloHitList().FillCaloHitList(deltaRayHitList);
+
+    CartesianPointVector deltaRayProjectedPositions;
+
+    for (const CaloHit *const pCaloHit : deltaRayHitList)
+        deltaRayProjectedPositions.push_back(pCaloHit->GetPositionVector());
+
+    return this->ParameteriseMuon(pParentMuon, deltaRayProjectedPositions, LArClusterHelper::GetClusterHitType(pDeltaRayCluster), positionOnMuon, muonDirection);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template<typename T>
+StatusCode NViewDeltaRayMatchingAlgorithm<T>::ParameteriseMuon(const ParticleFlowObject *const pParentMuon, const CartesianPointVector &deltaRayProjectedPositions,
+    const HitType hitType, CartesianVector &positionOnMuon, CartesianVector &muonDirection) const
+{
+    ClusterList muonClusterList;
+    LArPfoHelper::GetClusters(pParentMuon, hitType, muonClusterList);
+            
+    if (muonClusterList.size() != 1)
+        return STATUS_CODE_NOT_FOUND;
+    
+    CartesianPointVector muonProjectedPositions;
+    if (this->ProjectMuonPositions(hitType, pParentMuon, muonProjectedPositions) != STATUS_CODE_SUCCESS)
+        return STATUS_CODE_NOT_FOUND;
+
+    const Cluster *const pMuonCluster(muonClusterList.front());
+    const float slidingFitPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    const TwoDSlidingFitResult slidingFitResult(pMuonCluster, 40, slidingFitPitch);
+
+    CartesianVector deltaRayVertex(0.f, 0.f, 0.f), muonVertex(0.f, 0.f, 0.f);
+    LArMuonLeadingHelper::GetClosestPositions(deltaRayProjectedPositions, pMuonCluster, deltaRayVertex, muonVertex);
+
+    positionOnMuon = LArMuonLeadingHelper::GetClosestPosition(muonVertex, muonProjectedPositions, pMuonCluster);
+
+    if (positionOnMuon.GetMagnitude() < std::numeric_limits<float>::epsilon())
+        return STATUS_CODE_NOT_FOUND;
+
+    float rL(0.f), rT(0.f);
+    slidingFitResult.GetLocalPosition(positionOnMuon, rL, rT);
+    slidingFitResult.GetGlobalFitDirection(rL, muonDirection);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template<typename T>
```

Comment:
Probably worth moving this one to immediately below the alternative CollectHitsFromMuon function